Physical_Engine/Query/IsWithinPileCap.cs

```diff
@@ -0,0 +1,129 @@
+ * the Free Software Foundation, either version 3.0 of the License, or          
+ * (at your option) any later version.                                          
+ *                                                                              
+ * The BHoM is distributed in the hope that it will be useful,              
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of               
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the                 
+ * GNU Lesser General Public License for more details.                          
+ *                                                                            
+ * You should have received a copy of the GNU Lesser General Public License     
+ * along with this code. If not, see <https://www.gnu.org/licenses/lgpl-3.0.html>.      
+ */
+
+using System;
+using System.Linq;
+using System.Collections.Generic;
+using System.ComponentModel;
+using BH.oM.Physical.Reinforcement;
+using BH.oM.Quantities.Attributes;
+using BH.oM.Base.Attributes;
+using BH.oM.Physical.Elements;
+using BH.oM.Physical.FramingProperties;
+using BH.Engine.Spatial;
+using BH.oM.Geometry;
+using BH.Engine.Base;
+using BH.Engine.Geometry;
+
+namespace BH.Engine.Physical
+{
+    public static partial class Query
+    {
+        /***************************************************/
+        /**** Public Methods                            ****/
+        /***************************************************/
+
+        [Description("Determines whether the Piles are within the extents defined by the PadFoundation (both depth and extents). Piles located on the boundary are accepted within tolerance" +
+            "if their profile does not intersect with the pad foundation boundary.")]
+        [Input("padFoundation", "The PadFoundation that defines the extents for the Piles to be located.")]
+        [Input("piles", "The Piles to check if they are located within the extents of the PadFoundation.")]
+        [Input("tolerance", "Distance tolerance to be used in the method. Piles are deemed to be on the edge of the PadFoundation if they are within this distance from the curve.", typeof(Length))]
+        [Output("isWithinPileCap", "True if all Piles are contained within the boundary and depth of the PileCap. False if one or more Piles are not.")]
+        public static bool IsWithinPileCap(this PadFoundation padFoundation, List<Pile> piles, double tolerance = Tolerance.Distance)
+        {
+            if (padFoundation.IsNull())
+                return false;
+            else if (piles.IsNullOrEmpty())
+                return false;
+            else if (piles.Any(x => x.IsNull()))
+                return false;
+            else if (padFoundation.Construction.IThickness() == 0)
+            {
+                Base.Compute.RecordError("The PadFoundation is not assigned a thickness (within the Construction), the method will return false as it cannot determine the thickness" +
```

Comment:
As discussed the other day, I would try to avoid 'false' word
```suggestion
                Base.Compute.RecordError("The PadFoundation does not have thickness assigned (within the Construction), therefore piles are considered as outside of the pile cap." +
```